addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.openhab.binding.zoneminder.internal.state;
+
+import java.io.ByteArrayOutputStream;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.RawType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link GenericThingState} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Martin S. Eskildsen - Initial contribution
+ */
+public abstract class GenericThingState implements ChannelStateChangePublisher {
+
+    private Logger logger = LoggerFactory.getLogger(GenericThingState.class);
+
+    final AtomicBoolean allowRefresh = new AtomicBoolean(true);
+
+    private ChannelStateChangeSubscriber subscriber = null;
+
+    // Keeps track of current channel refresh status of subscribed channels
+    private Map<String, GenericChannelState> subscriptions = new ConcurrentHashMap<String, GenericChannelState>();
+
+    public GenericThingState(ChannelStateChangeSubscriber subscriber) {
+        this.subscriber = subscriber;
+    }
+
+    protected abstract void recalculate();
+
+    public void onChannelChanged(ChannelUID channelUID) {
+    }
+
+    @Override
+    public void addChannel(ChannelUID channelUID) {
+        try {
+            if (!subscriptions.containsKey(channelUID.getId())) {
+                subscriptions.put(channelUID.getId(), createChannelSubscription(channelUID));
+            }
+        } catch (
+
+        Exception ex) {
```

Comment:
Fix indentation and please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.